ledger/accountdb.go

```diff
@@ -4187,44 +4237,69 @@ func processAllBaseAccountRecords(
 	var addrbuf []byte
 	var buf []byte
 	var rowid int64
-	for baseRows.Next() {
-		err = baseRows.Scan(&rowid, &addrbuf, &buf)
-		if err != nil {
-			return 0, pendingRow{}, err
-		}
+	for {
+		if pendingBase.rowid != 0 {
+			addr = pendingBase.addr
+			rowid = pendingBase.rowid
+			accountData = *pendingBase.accountData
+			buf = pendingBase.encodedAccountData
+			pendingBase = pendingBaseRow{}
+		} else {
+			if !baseRows.Next() {
+				break
+			}
 
-		if len(addrbuf) != len(addr) {
-			err = fmt.Errorf("account DB address length mismatch: %d != %d", len(addrbuf), len(addr))
-			return 0, pendingRow{}, err
-		}
+			err = baseRows.Scan(&rowid, &addrbuf, &buf)
+			if err != nil {
+				return 0, pendingBaseRow{}, pendingResourceRow{}, err
+			}
 
-		copy(addr[:], addrbuf)
+			if len(addrbuf) != len(addr) {
+				err = fmt.Errorf("account DB address length mismatch: %d != %d", len(addrbuf), len(addr))
+				return 0, pendingBaseRow{}, pendingResourceRow{}, err
+			}
 
-		accountData = baseAccountData{}
-		err = protocol.Decode(buf, &accountData)
-		if err != nil {
-			return 0, pendingRow{}, err
+			copy(addr[:], addrbuf)
+
+			accountData = baseAccountData{}
+			err = protocol.Decode(buf, &accountData)
+			if err != nil {
+				return 0, pendingBaseRow{}, pendingResourceRow{}, err
+			}
 		}
+
 		err = baseCb(addr, rowid, &accountData, buf)
 		if err != nil {
-			return 0, pendingRow{}, err
+			return 0, pendingBaseRow{}, pendingResourceRow{}, err
 		}
 
-		pending, err = processAllResources(resRows, addr, &accountData, rowid, pending, resCb)
+		var moreRows bool
+		pendingResource, moreRows, err = processAllResources(resRows, addr, &accountData, rowid, pendingResource, resCb)
 		if err != nil {
 			err = fmt.Errorf("failed to gather resources for account %v, addrid %d, prev address %v : %w", addr, rowid, prevAddr, err)
-			return 0, pendingRow{}, err
+			return 0, pendingBaseRow{}, pendingResourceRow{}, err
+		}
+
+		if moreRows {
+			// we're done with this iteration.
```

Comment:
this looks inconsistent:
the original implementation supplies `accountCount` as a batch size, but for resources we limit it in resCb. I think it would be clearer and consistent if `processAllResources` would accept a batch size like this function. In this case we'll not need to circle the `moreRows` / `chunkFull` from the callback down to the readers and back.